src/Donations/Repositories/DonationRepository.php

```diff
@@ -0,0 +1,256 @@
+            )
+            ->where('post_type', 'give_payment')
+            ->whereIn('ID', function (QueryBuilder $builder) use ($donorId) {
+                $builder
+                    ->select('donation_id')
+                    ->from($this->donationMetaTable)
+                    ->where('meta_key', '_give_payment_donor_id')
+                    ->where('meta_value', $donorId);
+            })
+            ->orderBy('post_date', 'DESC')
+            ->getAll();
+
+        return array_map(static function ($donation) {
+            return DonationQueryData::fromObject($donation)->toDonation();
+        }, $donations);
+    }
+
+    /**
+     * @param  Donation  $donation
+     *
+     * @return Donation
+     * @throws Exception|DatabaseQueryException
+     */
+    public function insert(Donation $donation)
+    {
+        $date = current_datetime()->format('Y-m-d H:i:s');
+
+        DB::query('START TRANSACTION');
+
+        try {
+            DB::insert($this->postsTable, [
+                'post_date' => $date,
+                'post_date_gmt' => get_gmt_from_date($date),
+                'post_status' => $donation->status,
+                'post_type' => 'give_payment'
+            ], null);
+
+            $donationId = DB::last_insert_id();
+
+            foreach ($donation->getMeta() as $metaKey => $metaValue) {
+                DB::insert($this->donationMetaTable, [
+                    'donation_id' => $donationId,
+                    'meta_key' => $metaKey,
+                    'meta_value' => $metaValue,
+                ], null);
+            }
+        } catch (Exception $exception) {
+            DB::query('ROLLBACK');
+
+            Log::error('Failed creating a donation');
+
+            throw new $exception('Failed creating a donation');
+        }
+
+        DB::query('COMMIT');
+
+        return Donation::find($donationId);
+    }
+
+    /**
+     * @param  Donation  $donation
+     *
+     * @return Donation
+     * @throws Exception|DatabaseQueryException
+     */
+    public function update(Donation $donation)
+    {
+        $date = current_datetime()->format('Y-m-d H:i:s');
+
+        DB::query('START TRANSACTION');
+
+        try {
+            DB::update($this->postsTable, [
+                'post_modified' => $date,
+                'post_modified_gmt' => get_gmt_from_date($date),
+                'post_status' => $donation->status,
+                'post_type' => $donation->status
+            ], ['id' => $donation->id]);
+
+            foreach ($donation->getMeta() as $metaKey => $metaValue) {
+                DB::update($this->donationMetaTable, [
+                    'meta_value' => $metaValue,
+                ], [
+                    'donation_id' => $donation->id,
+                    'meta_key' => $metaKey
+                ]);
+            }
+        } catch (Exception $exception) {
+            DB::query('ROLLBACK');
+
+            Log::error('Failed creating a donation');
```

Comment:
@alaca  is the message incorrect, I think you want to say `updating` instead of creating?

Reply:
Good catch!